web/ww.ts

```diff
@@ -135,215 +270,106 @@ class Wormhole {
+		if (!id) {
+			return "";
+		}
 
-				const msg: RTCSessionDescriptionInit | null = JSON.parse(webwormhole.open(this.key, m.data));
-				if (!msg) {
-					this.fail("bad key");
-					this.ws.send(webwormhole.seal(this.key, "bye"));
-					this.ws.close(WormholeErrorCodes.closeBadKey);
-					return;
-				}
-				if (msg.type !== "answer") {
-					console.log("unexpected message", msg);
-					this.fail("unexpected message");
-					return;
-				}
-				console.log("got answer");
-				this.pc.setRemoteDescription(new RTCSessionDescription(msg));
-				this.resolve3(webwormhole.fingerprint(this.key));
-				this.state = "wait_for_candidates";
-				return;
+		let conntype: string = "";
+		stats.forEach((s) => {
+			if (s.id === id) {
+				conntype = (s as { candidateType: string }).candidateType;
 			}
+		});
 
-			case "wait_for_candidates": {
-				if (!this.key || !this.pc) { return }
+		return conntype;
+	}
 
-				const msg: {candidate: string} | null = JSON.parse(webwormhole.open(this.key, m.data));
-				if (!msg) {
-					this.fail("bad key");
-					this.ws.send(webwormhole.seal(this.key, "bye"));
-					this.ws.close(WormholeErrorCodes.closeBadKey);
-					return;
-				}
-				console.log("got remote candidate", msg.candidate);
-				this.promise2.then(async () => {
-					if (!this.key || !this.pc) { return }
-					this.pc.addIceCandidate(new RTCIceCandidate(msg));
-				});
-				return;
-			}
+	async dial(): Promise<Uint8Array> {
+		this.ws = new WebSocket(
+			Wormhole.wsserver(this.signalserver, this.slot),
+			Wormhole.protocol
+		);
+		// Use lambdas so that 'this' in the respective bodies refers to the Wormhole
+		// instance, and not the WebSocket one.
+		this.ws.onopen = () => this.onopen();
+		this.ws.onerror = (e: Event) => this.onerror(e);
+		this.ws.onclose = (e: CloseEvent) => this.onclose(e);
+		this.ws.onmessage = (e: MessageEvent) => this.onmessage(e);
 
-			case "wait_for_pc_initialize":
-			case "wait_for_local_offer":
-			case "wait_for_local_answer": {
-				console.log("unexpected message", m);
-				this.fail("unexpected message");
-				return;
-			}
-			case "error":
-				return;
-		}
+		return this.done;
 	}
 
-	makePeerConnection(iceServers: GoICEServers) {
-		let normalisedICEServers = [];
-		for (let i = 0; i < iceServers.length; i++) {
-			normalisedICEServers.push({
-				urls: iceServers[i].URLs,
-				username: iceServers[i].Username,
-				credential: iceServers[i].Credential,
-			});
+	async onmessage(m: MessageEvent) {
+		if (!this.ws) {
+			return;
 		}
-		this.pc = new RTCPeerConnection({
-			iceServers: normalisedICEServers,
-		});
-		this.pc.onicecandidate = (e: RTCPeerConnectionIceEvent) => {
-			if (e.candidate && e.candidate.candidate !== "") {
-				console.log("got local candidate", e.candidate.candidate);
-				if (this.key) {
-					this.ws.send(webwormhole.seal(this.key, JSON.stringify(e.candidate)));
-				}
-			}
-		};
+		// Feed the state machine a new message.
+		this.state = await this.state(m.data);
```

Comment:
This new method of transitioning through the state machine is interesting. I like how it doesn't require that you have to remember to explicitly update the current state. It's also nice that it couples every message with a state transition.

I do think however that the states and the transitions are somewhat hidden to the reader of this code. It might be clearer to follow a more declarative approach, [something like this](https://xstate.js.org/viz/?gist=0a97c0512721ec401dd40d6be90fc675). ie list all the states and transitions up front. Another drawback to the current approach is you can't see what the current state is.